File: day05/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"slices"
	"strconv"
	"strings"

	"github.com/xorkevin/advent2024/bitset"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	log.SetFlags(log.Lshortfile)

	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	count1 := 0
	count2 := 0
	orders := [100][100]byte{}
	seen := bitset.New(100)

	firstHalf := true
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		if firstHalf {
			if len(scanner.Bytes()) == 0 {
				firstHalf = false
				continue
			}
			f := strings.Split(scanner.Text(), "|")
			num1, err := strconv.Atoi(f[0])
			if err != nil {
				log.Fatalln(err)
			}
			num2, err := strconv.Atoi(f[1])
			if err != nil {
				log.Fatalln(err)
			}
			// num1 must be ordered before num2
			current := orders[num1][0] + 1
			orders[num1][current] = byte(num2)
			orders[num1][0] = current
			continue
		}
		f := strings.Split(scanner.Text(), ",")
		nums := make([]byte, 0, len(f))
		for _, i := range f {
			num, err := strconv.Atoi(i)
			if err != nil {
				log.Fatalln(err)
			}
			nums = append(nums, byte(num))
		}

		inOrder := true
		seen.Reset()
	outer:
		for _, i := range nums {
			for _, o := range orders[i][1 : orders[i][0]+1] {
				if seen.Contains(int(o)) {
					inOrder = false
					break outer
				}
			}
			seen.Insert(int(i))
		}
		if inOrder {
			count1 += int(nums[len(nums)/2])
		} else {
			slices.SortFunc(nums, func(a, b byte) int {
				if a == b {
					return 0
				}
				for _, i := range orders[a][1 : orders[a][0]+1] {
					if i == b {
						return -1
					}
				}
				for _, i := range orders[b][1 : orders[b][0]+1] {
					if i == a {
						return 1
					}
				}
				return 0
			})
			count2 += int(nums[len(nums)/2])
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", count1)
	fmt.Println("Part 2:", count2)
}

File: day05/src/main.rs
use advent2024::bitset::BitSet;
use std::cmp::Ordering;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut count1 = 0;
    let mut count2 = 0;
    let mut orders = [[0 as u8; 100]; 100];
    let mut seen = BitSet::new(100);

    let mut first_half = true;
    for line in reader.lines() {
        let line = line?;
        if first_half {
            if line.len() == 0 {
                first_half = false;
                continue;
            }
            let mut f = line.splitn(2, "|");
            let num1 = f.next().ok_or("Missing first num")?.parse::<u8>()?;
            let num2 = f.next().ok_or("Missing second num")?.parse::<u8>()?;
            let current = orders[num1 as usize][0] + 1;
            orders[num1 as usize][current as usize] = num2;
            orders[num1 as usize][0] = current;
            continue;
        }
        let mut in_order = true;
        let mut nums = line
            .split(",")
            .map(|v| v.parse::<u8>())
            .collect::<Result<Vec<_>, _>>()?;
        seen.reset();
        'outer: for &i in nums.iter() {
            for &o in &orders[i as usize][1..((orders[i as usize][0] + 1) as usize)] {
                if seen.contains(o as usize) {
                    in_order = false;
                    break 'outer;
                }
                seen.insert(i as usize);
            }
        }
        if in_order {
            count1 += nums[nums.len() / 2] as u32;
        } else {
            nums.sort_by(|&a, &b| {
                if a == b {
                    return Ordering::Equal;
                }
                for &i in &orders[a as usize][1..((orders[a as usize][0] + 1) as usize)] {
                    if i == b {
                        return Ordering::Less;
                    }
                }
                for &i in &orders[b as usize][1..((orders[b as usize][0] + 1) as usize)] {
                    if i == a {
                        return Ordering::Greater;
                    }
                }
                Ordering::Equal
            });
            count2 += nums[nums.len() / 2] as u32;
        }
    }

    println!("Part 1: {}", count1);
    println!("Part 2: {}", count2);
    Ok(())
}
